### DISCUSSION TITLE
Understanding aln-r2 output

### DISCUSSION THREAD

**Discussion by Pkaps25** [ChaissonLab/danbing-tk#5] [Show and tell]
Hi,

Attached is a `*.rawLR.pred` file from one of the samples with SRS data I used to create my RPGG. Maybe I am not fully understanding this output, but if the ground truth k-mer counts are identical to the SRS k-mer counts at a given locus, wouldn't we expect the corresponding r^2 value to be closer to 1? In the cases where the k-mer counts are equal, the r^2 value is always 0.

![Screen Shot 2021-01-20 at 10 28 03 AM](https://user-images.githubusercontent.com/43655728/105197112-46c44300-5b0a-11eb-9a2a-9ccf9b8e6cb2.png)

**Comment by joyeuxnoel8** [comment]
Ah right, this is because all the kmer counts in the assembly e.g. `tb_south_am` have the same value. The output r^2 will always be 0 in this case. Nonetheless, it's possible to prevent this by adding a dummy variable at (0,0) since the intercept term should always be 0 in this context. I'll update the code shortly so that it makes more sense to have a value closer to 1 instead of 0.

**Comment by joyeuxnoel8** [comment]
`script/kmers.linreg.py` is now updated!

**Reply by Pkaps25** [reply]
Great! Thank you so much for your help. I would like to ask about the result of aligning SRS reads to an RPGG. Could you explain how to interpret the `*.tr.kmers` file that results from aligning to the graph? Is it possible to use this file directly to differentiate the VNTR composition of samples, perhaps by comparing the kmer counts at the various VNTR loci?

**Reply by joyeuxnoel8** [reply]
Comparing the kmer counts at the same locus across samples are valid when:
- If SRS datasets are coming from the same batch and sequencing technology --> valid
- else
  - if SRS datasets have similar sampling biases (check [here](https://github.com/ChaissonLab/danbing-tk#distribution-of-lsb) for details) --> valid
  - else --> invalid

Comparing the kmer counts between loci across samples are most likely invalid, unless the two loci have similar sampling biases.

When analyzing the 1KGP and GTEx samples in the manuscript, we also had to impose this constraint and remove outliers with distinct sampling biases.

**Comment by joyeuxnoel8** [comment]
It seems that you have sorted things out. But just to make sure the usage
is clear: the leave-one-out (LOO) analysis is meant to evaluate the
performance/quality of an RPGG by checking how accurate the length
estimates are, assuming that the set of SRS samples used have similar LSBs.
The LSB analysis allows us to evaluate whether the assumption holds, which
is a necessary requirement for length prediction and therefore LOO analysis.

And yes the bed file for non-repetitive regions is also released under v1.0
<https://github.com/ChaissonLab/danbing-tk/releases/tag/v1.0>.

**Reply by Pkaps25** [reply]
Hello, I have expanded the size of the VNTR set I pass in to RPGG construction and the build now fails. After exceeding ~100 regions in the TR bed file, the build errors. Snakemake log attached.
[2021-02-08T142910.591991.snakemake.log](https://github.com/ChaissonLab/danbing-tk/files/5946422/2021-02-08T142910.591991.snakemake.log)

**Reply by joyeuxnoel8** [reply]
Hello, could you also show the stderr printed to the command prompt besides the snakemake log? Snakemake logs only indicate the step that failed. Error messages from the failed command would be printed to the command prompt if running locally, or written to the log file for that particular job if submitting to the cluster.

**Reply by Pkaps25** [reply]
Of course, apologies for forgetting to include it. Here is the error:

![Screen Shot 2021-02-08 at 4 17 32 PM](https://user-images.githubusercontent.com/43655728/107282717-5cea6280-6a29-11eb-99f3-114d15454d3a.png)

**Reply by joyeuxnoel8** [reply]
No worries! Thanks for providing the info. I just pushed a change to help debug. For some reason the `pan.tr.mbe.v0.bed` indicates a different number of supports from `{g}/tmp1.{h}.bed`.  Could you run `JointTRAnnotation` again and show the two files with discrepancy? I suspect that the problem may have occurred during liftover.

**Reply by Pkaps25** [reply]
It appears the file `script/preMBE.py` has been removed. Was this intentional?

**Reply by joyeuxnoel8** [reply]
Yes, I merged a few redundant operations and merged them in `script/multiBoundaryExpansion.py`

**Reply by Pkaps25** [reply]
Here is the trace from the latest run
<img width="1217" alt="Screen Shot 2021-02-08 at 9 37 52 PM" src="https://user-images.githubusercontent.com/43655728/107308071-f4fe4100-6a55-11eb-9e81-fabb42aeeb77.png">

These are the inconsistent files:
[pan.tr.mbe.v0.txt](https://github.com/ChaissonLab/danbing-tk/files/5948387/pan.tr.mbe.v0.txt)
[tmp1.0.txt](https://github.com/ChaissonLab/danbing-tk/files/5948390/tmp1.0.txt)

Also, in lines 268 and 269 of `GoodPangenomeGraph.snakefile`, bed columns are expanded by 700 bp. `params.FS` is used as input to scripts, but here the 700bp flank is hardcoded. Is this intended?

**Reply by joyeuxnoel8** [reply]
I noticed that several regions in `tr.regions.bed` overlap with each other, e.g. 

> NC000913.3      59178   59184
> NC000913.3      59178   59230

The provided reference regions must be non-overlapping. I assume that this is a test run since those regions don't really count as VNTRs. In case you're running on real VNTR regions, I would recommend using `bedtools merge -d {FS}` to ensure the reference VNTR regions are not too close to each other.

**Reply by joyeuxnoel8** [reply]
Also, `params.FS` is now the size of flanking sequences. Thanks for the catch!

**Reply by Pkaps25** [reply]
Thank you, running successfully now.

**Reply by Pkaps25** [reply]
Hi, I am receiving a different error in the `JointTRAnnotation` rule, I think when merging regions across haplotypes. I have merged overlapping TR regions prior to running and have attached the trace. Let me know what other information may be helpful. 
[15klog.txt](https://github.com/ChaissonLab/danbing-tk/files/6015134/15klog.txt)

**Reply by Pkaps25** [reply]
Resolved: Problem was using an old Snakefile where FS was hardcoded and did not match the FS in config file.

**Reply by joyeuxnoel8** [reply]
Nice, thanks for the update!
